lib/src/concat.dart

```diff
@@ -17,11 +20,66 @@ class _Concat<T> implements StreamTransformer<T, T> {
 
   @override
   Stream<T> bind(Stream<T> first) {
-    var controller = new StreamController<T>();
-    controller
-        .addStream(first)
-        .then((_) => controller.addStream(_next))
-        .then((_) => controller.close());
+    var controller = first.isBroadcast
+        ? new StreamController<T>.broadcast(sync: true)
+        : new StreamController<T>(sync: true);
+
+    StreamSubscription subscription;
+    var currentStream = first;
+    var firstDone = false;
+    var secondDone = false;
+
+    Function currentDoneHandler;
+
+    listen() {
+      subscription = currentStream.listen(controller.add,
+          onError: controller.addError, onDone: () => currentDoneHandler());
+    }
+
+    onSecondDone() {
+      secondDone = true;
+      controller.close();
+    }
+
+    onFirstDone() {
+      firstDone = true;
+      currentStream = _next;
+      currentDoneHandler = onSecondDone;
+      listen();
+    }
+
+    currentDoneHandler = onFirstDone;
+
+    controller.onListen = () {
+      if (firstDone &&
```

Comment:
I am a bit confused by a lot of the logic here, why do we only care if `_next` is a broadcast stream, shouldn't most of these be checking for `currentStream.isBroadcast`?